aptos-move/framework/aptos-framework/sources/voting.move

```diff
@@ -326,6 +373,67 @@ module aptos_framework::voting {
         option::extract(&mut proposal.execution_content)
     }
 
+    /// Resolve a single-step or a multi-step proposal with given id.
+    /// Can only be done if there are at least as many votes as min required and
+    /// there are more yes votes than no. If either of these conditions is not met, this will revert.
+    ///
+    /// @param voting_forum_address The address of the forum where the proposals are stored.
+    /// @param proposal_id The proposal id.
+    /// @param next_execution_hash The next execution hash if the given proposal is multi-step.
+    public fun resolve_single_step_or_multi_step_proposal<ProposalType: store>(
+        voting_forum_address: address,
+        proposal_id: u64,
+        next_execution_hash: vector<u8>,
+    ) acquires VotingForum {
+        let proposal_state = get_proposal_state<ProposalType>(voting_forum_address, proposal_id);
+        assert!(proposal_state == PROPOSAL_STATE_SUCCEEDED, error::invalid_state(EPROPOSAL_CANNOT_BE_RESOLVED));
+
+        let voting_forum = borrow_global_mut<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = table::borrow_mut(&mut voting_forum.proposals, proposal_id);
+        assert!(!proposal.is_resolved, error::invalid_state(EPROPOSAL_ALREADY_RESOLVED));
+
+        // If this proposal is multi-step, we update the metadata map to indicate that this proposal is in execution.
+        let multi_step_in_execution_key = utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        if (simple_map::contains_key(&proposal.metadata, &multi_step_in_execution_key)) {
+            let is_multi_step_proposal_in_execution_value = simple_map::borrow_mut(&mut proposal.metadata, &multi_step_in_execution_key);
+            if (!from_bcs::to_bool(*is_multi_step_proposal_in_execution_value)) {
+                *is_multi_step_proposal_in_execution_value = to_bytes(&true);
+            };
+        };
+
+        let resolvable_time = to_u64(*simple_map::borrow(&proposal.metadata, &utf8(RESOLVABLE_TIME_METADATA_KEY)));
```

Comment:
nit: This seems redundant. I wonder if we can rollup a lot of these checks around whether the proposal can be resolved so we don't have to duplicate them among the single-step and multi-step proposal execution functions

Reply:
yeah, or we can just direct the existing single-step resolve function to call the newly added generic resolve function.  

the checks are mostly the same, it's just the event-emitting logic and updating to next execution hash logic are different. I can add a new function `check_if_proposal_is_resolvable` to do all the checks as well, but it'll be trickier, since we're borrowing from `VotingForum` to emit events, but we're also going to use `VotingForum` in the `check_if_proposal_is_resolvable()` function as well, and I don't think that's allowed so we will probably need to do something hacky.

Reply:
I see that's probably fine then.

Reply:
good point! I just updated this to include a `is_proposal_resolvable()` function to do all the checks because I realized that as long as we are not borrowing `VotingForum` before calling `is_proposal_resolvable()`, we will be fine.